File: csharp/lib/SequenceSolver.cs
namespace ClounceMath {
    using System;

    public sealed class SequenceSolver {

        public static void ArithmeticMeanSequence(double sequenceStartValue, double sequenceEndValue, ref double[] variables)
        {
            int numberOfVariables = variables == null ? 0 : variables.Length;

            // If there are no variables then skip computation
            if (variables == null || numberOfVariables == 0) {
                return;
            }

            double sequenceNumber = Sequences.JacobsthalNumberNonRecursive(numberOfVariables);
            double nextSequenceNumber = Sequences.JacobsthalNumberNonRecursive(numberOfVariables + 1);
            variables[0] = ((Math.Pow(2, numberOfVariables) * sequenceEndValue) - (sequenceNumber * sequenceStartValue)) / nextSequenceNumber;

            for (int i = 1; i < numberOfVariables; i++) {
                sequenceNumber = Sequences.JacobsthalNumberNonRecursive(i);
                nextSequenceNumber = Sequences.JacobsthalNumberNonRecursive(i + 1);
                variables[i] = ((sequenceNumber * sequenceStartValue) + (nextSequenceNumber * variables[0])) / Math.Pow(2, i);
            }
        }

        public static double ArithmeticMeanSequenceVariable(double sequenceStartValue, double sequenceEndValue, int numberOfVariables, int variableToSolve)
        {
            // If there are no variables then skip computation
            if (numberOfVariables == 0 || variableToSolve > numberOfVariables) {
                return 0;
            }

            double nthJacobsthal = Sequences.JacobsthalNumberNonRecursive(variableToSolve);
            double preNthJacobsthal = Sequences.JacobsthalNumberNonRecursive(variableToSolve - 1);
            double sizeVarJacobsthal = Sequences.JacobsthalNumberNonRecursive(numberOfVariables);
            double nextSizeVarJacobsthal = Sequences.JacobsthalNumberNonRecursive(numberOfVariables + 1);

            double yCalculation = Math.Pow(2, numberOfVariables) * nthJacobsthal * sequenceEndValue;
            double xCalculation = ((Math.Pow(-1, variableToSolve) * sizeVarJacobsthal) +
                (Math.Pow(-1, numberOfVariables) * preNthJacobsthal)) * sequenceStartValue;
            double numerator = yCalculation + xCalculation;
            double denominator = nextSizeVarJacobsthal * Math.Pow(2, variableToSolve - 1);

            return numerator / denominator;
        }
    }
}

File: python/lib/sequence_solver.py
import math
from sequences import Sequences

class SequenceSolver:
    @staticmethod
    def arithmetic_mean_sequence(sequence_start_value, sequence_end_value, variables):
        number_of_variables = len(variables)

        if (variables is None or number_of_variables == 0) :
            return 0
        
        sequence_number = Sequences.jacobsthal_number_non_recursive(number_of_variables)
        next_sequence_number = Sequences.jacobsthal_number_non_recursive(number_of_variables + 1)

        variables[0] = (((2 ** number_of_variables) * sequence_end_value) - (sequence_number * sequence_start_value)) / next_sequence_number

        for i in range(1, number_of_variables):
            sequence_number = Sequences.jacobsthal_number_non_recursive(i)
            next_sequence_number = Sequences.jacobsthal_number_non_recursive(i + 1)
            variables[i] = ((sequence_number * sequence_start_value) + (next_sequence_number * variables[0])) / 2 ** i
    
    @staticmethod
    def arithmetic_mean_sequence_variable(sequence_start_value, sequence_end_value, number_of_variables, variable_to_solve):
        if (number_of_variables == 0 or variable_to_solve > number_of_variables):
            return 0
        
        nth_jacobsthal = Sequences.jacobsthal_number_non_recursive(variable_to_solve)
        pre_nth_jacobsthal = Sequences.jacobsthal_number_non_recursive(variable_to_solve - 1)
        size_var_jacobsthal = Sequences.jacobsthal_number_non_recursive(number_of_variables)
        next_size_var_jacobsthal = Sequences.jacobsthal_number_non_recursive(number_of_variables + 1)
        
        y_calculation = (2  ** number_of_variables) * nth_jacobsthal * sequence_end_value
        x_calculation = ((pow(-1, variable_to_solve) * size_var_jacobsthal) \
            + (pow(-1, number_of_variables) * pre_nth_jacobsthal)) * sequence_start_value
        numerator = y_calculation + x_calculation
        denominator = next_size_var_jacobsthal * (2 ** (variable_to_solve - 1))

        return numerator / denominator